flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java

```diff
@@ -0,0 +1,117 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptive;
+
+import org.apache.flink.runtime.checkpoint.CompletedCheckpoint;
+import org.apache.flink.runtime.checkpoint.StopWithSavepointOperations;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.Execution;
+import org.apache.flink.runtime.executiongraph.ExecutionGraph;
+import org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointOperationHandler;
+import org.apache.flink.runtime.scheduler.stopwithsavepoint.StopWithSavepointOperationHandlerImpl;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.stream.Collectors;
+
+/**
+ * A thin wrapper around the StopWithSavepointOperationHandler to handle global failures according
+ * to the needs of AdaptiveScheduler. AdaptiveScheduler currently doesn't support local failover,
+ * hence, any reported failure will lead to a transition to Restarting or Failing state.
+ */
+class StopWithSavepointOperationManagerForAdaptiveScheduler implements GlobalFailureHandler {
+    private final StopWithSavepointOperationHandler handler;
+    private final ExecutionGraph executionGraph;
+    private final StopWithSavepointOperations stopWithSavepointOperations;
+    private boolean operationFinished = false;
+
+    StopWithSavepointOperationManagerForAdaptiveScheduler(
+            ExecutionGraph executionGraph,
+            StopWithSavepointOperations stopWithSavepointOperations,
+            boolean terminate,
+            @Nullable String targetDirectory,
+            Executor mainThreadExecutor,
+            Logger logger) {
+
+        this.stopWithSavepointOperations = stopWithSavepointOperations;
+
+        // do not trigger checkpoints while creating the final savepoint. We will start the
+        // scheduler onLeave() again.
+        stopWithSavepointOperations.stopCheckpointScheduler();
+
+        final CompletableFuture<CompletedCheckpoint> savepointFuture =
+                stopWithSavepointOperations.triggerSynchronousSavepoint(terminate, targetDirectory);
+
+        this.handler =
+                new StopWithSavepointOperationHandlerImpl(
+                        executionGraph.getJobID(), this, stopWithSavepointOperations, logger);
+
+        FutureUtils.assertNoException(
+                savepointFuture
+                        // the completedSavepointFuture could also be completed by
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
+                        .handleAsync(
+                        (completedSavepoint, throwable) -> {
+                            handler.handleSavepointCreation(completedSavepoint, throwable);
```

Comment:
I think we will stay forever in the state `StopWithSavepoint` if the savepoint future fails.